apple/Tests/FerrostarCoreTests/Models/CoreLocationModelTests.swift

```diff
@@ -0,0 +1,103 @@
+import XCTest
+import CoreLocation
+import UniFFI
+@testable import FerrostarCore
+
+class MockCLHeading: CLHeading {
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    init(trueHeading: CLLocationDirection,
+         headingAccuracy: CLLocationDirectionAccuracy,
+         timestamp: Date) {
+        
+        self._trueHeading = trueHeading
+        self._headingAccuracy = headingAccuracy
+        self._timestamp = timestamp
+        super.init()
+    }
+    
+    let _trueHeading: CLLocationDirection
+    override var trueHeading: CLLocationDirection {
+        _trueHeading
+    }
+    
+    let _headingAccuracy: CLLocationDirectionAccuracy
+    override var headingAccuracy: CLLocationDirection {
+        _headingAccuracy
+    }
+    
+    let _timestamp: Date
+    override var timestamp: Date {
+        _timestamp
+    }
+}
+
+final class CoreLocationModelTests: XCTestCase {
+    
+    // MARK: UserLocation
+    
+    func testInitUserLocation() {
+        let timestamp = Date()
+        let coordinate = CLLocationCoordinate2D(latitude: -77.846323, longitude: 166.668235)
+        let location = CLLocation(coordinate: coordinate,
+                                  altitude: 10,
+                                  horizontalAccuracy: 1.1,
+                                  verticalAccuracy: 6.6,
+                                  course: 45.5,
+                                  courseAccuracy: 3.3,
+                                  speed: 4.4,
+                                  speedAccuracy: 1.1,
+                                  timestamp: timestamp)
+        
+        let userLocation = UserLocation(clLocation: location)
+        
+        XCTAssertEqual(userLocation.coordinates.lat, -77.846323)
+        XCTAssertEqual(userLocation.coordinates.lng, 166.668235)
+        XCTAssertEqual(userLocation.horizontalAccuracy, 1.1)
+        XCTAssertEqual(userLocation.courseOverGround?.degrees, 45)
+        XCTAssertEqual(userLocation.courseOverGround?.accuracy, 3)
+        XCTAssertEqual(userLocation.timestamp, timestamp)
+    }
+    
+    // TODO: Decide methodology for invalid CLLocation data (negatives)
```

Comment:
Good question. Off the top of my head, I'd say the following are invalid:

* Lat > 90
* Lat < -90
* Lng > 180
* Lng < 180
* Negative heading
* Negative course
* Negative accuracy (on any model that includes accuracy)

Reply:
Let chat about this item. After having played around with CLLocation to see what it'll do I'm less certain we want to front load this in a convenience init. E.g. we might find a slightly invalid location is better for navigation than just discarding it in full. I'm also curious how bad a system provided location can get.

Reply:
I'm not quite following what you mean by a "slightly invalid location." A location is either valid or invalid. There isn't a middle ground.

Whether a CLLocation is useful at a given _accuracy_ is tunable by the user already at multiple levels of Ferrostar. However, any CLLocation with a negative accuracy is explicitly invalid and coordinates out of the range cannot be reliably interpreted so they should be discarded (we could enforce this in Rust; these will never be valid).

CLLocation reference on the accuracy field (we should discard all values with negative accuracy): https://developer.apple.com/documentation/corelocation/cllocation/1423599-horizontalaccuracy